From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Owen <23108066+Owen1212055@users.noreply.github.com>
Date: Fri, 20 Jun 2025 18:22:37 -0400
Subject: [PATCH] Paper: Cleanup Primed TNT Fix

Original license: GPLv3
Original project: https://github.com/PaperMC/Paper

https://github.com/PaperMC/Paper/commit/6a51c44ec28830094de58528fd8107c4be5cefe0

This option should probably be removed as its a descendant of an option that allows "old tnt cannon" behavior before ~1.9
But this improves the fix so properly update the velocity/position rather than the questionable way it was doing it before.

diff --git a/net/minecraft/world/entity/item/PrimedTnt.java b/net/minecraft/world/entity/item/PrimedTnt.java
index 8aedc3ca463745fe32cac977208b23dc0b8e73b6..2b12316a7e8f26a7d09e3de1da57166b02a08a8a 100644
--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -135,25 +135,8 @@ public class PrimedTnt extends Entity implements TraceableEntity {
         }
         // Paper start - Option to prevent TNT from moving in water
         if (!this.isRemoved() && this.wasTouchingWater && this.level().paperConfig().fixes.preventTntFromMovingInWater) {
-            /*
-             * Author: Jedediah Smith <jedediah@silencegreys.com>
-             */
-            // Send position and velocity updates to nearby players on every tick while the TNT is in water.
-            // This does pretty well at keeping their clients in sync with the server.
-            net.minecraft.server.level.ChunkMap.TrackedEntity ete = ((net.minecraft.server.level.ServerLevel) this.level()).getChunkSource().chunkMap.entityMap.get(this.getId());
-            if (ete != null) {
-                net.minecraft.network.protocol.game.ClientboundSetEntityMotionPacket velocityPacket = new net.minecraft.network.protocol.game.ClientboundSetEntityMotionPacket(this);
-                net.minecraft.network.protocol.game.ClientboundTeleportEntityPacket positionPacket = net.minecraft.network.protocol.game.ClientboundTeleportEntityPacket.teleport(this.getId(), net.minecraft.world.entity.PositionMoveRotation.of(this), java.util.Set.of(), this.onGround);
-
-                // Leaf start - Multithreaded tracker
-                for (var viewer : ete.seenBy()) {
-                    if ((viewer.getPlayer().getX() - this.getX()) * (viewer.getPlayer().getY() - this.getY()) * (viewer.getPlayer().getZ() - this.getZ()) < 16 * 16) {
-                        viewer.send(velocityPacket);
-                        viewer.send(positionPacket);
-                    }
-                }
-                // Leaf end - Multithreaded tracker
-            }
+            this.hurtMarked = true; // Paper - Cleanup Primed TNT Fix
+            this.hasImpulse = true; // Paper - Cleanup Primed TNT Fix
         }
         // Paper end - Option to prevent TNT from moving in water
     }
